### PR TITLE
Reduce memory usage of scheduler process - Optimize scheduler.py::TaskState class

### PR DESCRIPTION
**Pull request by milesgranger**
Part of #7998 

The first approach in https://github.com/dask/distributed/pull/8331/commits/ef4b7ddb9fa25795bf43e0de2a8f6cd584f0f5ad moved things into descriptors, which would only assign a value when being accessed.  This results in ~34% reduction in memory when compared to main. Downside was there is some boilerplace for each attribute requiring a lazy creation (albeit this could be generalized by prefixing each with `__lazy_` for example then iterating on those annotations after initialization..)

Second approach, the current direction of this PR, uses optionals for more attributes in this class. Primarily those using `set`/`dict`. This results in ~42% reduction in memory when compared to main. Downside is much more changed code, checking for `None` everywhere and the tests are currently wildly broken. :sweat_smile: 

Snippet I'm using for this comparison to main:
```python
from dask.datasets import timeseries
from dask.distributed import Client
from distributed.diagnostics.memray import memray_scheduler

def main():
    ddf = timeseries("2020", "2021", partition_freq='2h')
    ddf2 = ddf.shuffle(on="x", shuffle='tasks')

    print(len(ddf2.__dask_graph__()))

    with Client() as client:
        with memray_scheduler():
            fut = client.compute(ddf2.size)
            import time
            time.sleep(90)
        fut.cancel()

if __name__ == '__main__':
    main()

```
cc @fjetter

### CODE REVIEW

**Comment by milesgranger** [comment]
Okay, @fjetter I think this is ready for a proper review then. Failing tests don't seem related, (ran most/all of those failing here locally fine as well).

File: distributed/scheduler.py [hunk 1]
```diff
@@ -782,18 +782,19 @@ def _remove_from_task_prefix_count(self, ts: TaskState) -> None:
     def remove_replica(self, ts: TaskState) -> None:
         """The worker no longer has a task in memory"""
         if self.scheduler.validate:
-            assert self in ts.who_has
+            assert self in (ts.who_has or ())
```
Comment (fjetter) [inline distributed/scheduler.py:785]:
```suggestion
            assert ts.who_has
            assert self in ts.who_has
```

File: distributed/scheduler.py [hunk 2]
```diff
@@ -782,18 +782,19 @@ def _remove_from_task_prefix_count(self, ts: TaskState) -> None:
...
             assert ts in self.has_what
             assert ts not in self.needs_what
 
         self.nbytes -= ts.get_nbytes()
         del self._has_what[ts]
-        ts.who_has.remove(self)
+        if ts.who_has:
+            ts.who_has.remove(self)
```
Comment (fjetter) [inline distributed/scheduler.py:792]: The fact that this was a `remove` and not a discard already proves that this is a non-empty set
```suggestion
        ts.who_has.remove(self)
```

Comment (fjetter) [reply distributed/scheduler.py:792]: one could consider something like

```python
ts.who_has.remove(self)
if not ts.who_has:
    ts.who_has = None
```
but I'm not sure if that's worth it

File: distributed/scheduler.py [hunk 3]
```diff
@@ -1447,8 +1450,13 @@ def state(self, value: TaskStateState) -> None:
 
     def add_dependency(self, other: TaskState) -> None:
         """Add another task as a dependency of this task"""
+        if self.dependencies is None:
+            self.dependencies = set()
```
Comment (fjetter) [inline distributed/scheduler.py:1454]: As already pointed out above, there are very, very few cases where a Task doesn't walk through this code path. Therefore, I suggest to keep this strictly non-optional

File: distributed/scheduler.py [hunk 4]
```diff
@@ -1447,8 +1450,13 @@ def state(self, value: TaskStateState) -> None:
...
         self.dependencies.add(other)
         self.group.dependencies.add(other.group)
+
+        if other.dependents is None:
+            other.dependents = set()
```
Comment (fjetter) [inline distributed/scheduler.py:1459]: also non-optional

File: distributed/scheduler.py [hunk 5]
```diff
@@ -2548,12 +2563,14 @@ def transition_waiting_released(self, key: Key, stimulus_id: str) -> RecsMsgs:
...
-                dts.waiters.discard(ts)
+        for dts in ts.dependencies or ():
+            if ts in (dts.waiters or ()):
+                if dts.waiters:
+                    dts.waiters.discard(ts)
                 if not dts.waiters and not dts.who_wants:
                     recommendations[dts.key] = "released"
-        ts.waiting_on.clear()
+        if ts.waiting_on:
+            ts.waiting_on.clear()
```
Comment (fjetter) [inline distributed/scheduler.py:2573]: Instead of clearing we can probably reset it to `None`

File: distributed/scheduler.py [hunk 6]
```diff
@@ -2528,12 +2542,13 @@ def transition_erred_released(self, key: Key, stimulus_id: str) -> RecsMsgs:
             "keys": [key],
             "stimulus_id": stimulus_id,
         }
-        for ws_addr in ts.erred_on:
+        for ws_addr in ts.erred_on or ():
             worker_msgs[ws_addr] = [w_msg]
-        ts.erred_on.clear()
+        if ts.erred_on:
+            ts.erred_on.clear()
```
Comment (fjetter) [inline distributed/scheduler.py:2548]:
```suggestion
        ts.erred_on = None
```

File: distributed/scheduler.py [hunk 7]
```diff
@@ -2562,7 +2579,8 @@ def transition_waiting_released(self, key: Key, stimulus_id: str) -> RecsMsgs:
         elif not ts.exception_blame and (ts.who_wants or ts.waiters):
             recommendations[key] = "waiting"
         else:
-            ts.waiters.clear()
+            if ts.waiters:
+                ts.waiters.clear()
```
Comment (fjetter) [inline distributed/scheduler.py:2583]:
```suggestion
            ts.waiters = None
```

File: distributed/scheduler.py [hunk 8]
```diff
@@ -2669,16 +2689,18 @@ def transition_processing_erred(
...
-            dts.waiters.discard(ts)
+        for dts in ts.dependencies or ():
+            if dts.waiters:
+                dts.waiters.discard(ts)
             if not dts.waiters and not dts.who_wants:
                 recommendations[dts.key] = "released"
 
-        ts.waiters.clear()  # do anything with this?
+        if ts.waiters:
+            ts.waiters.clear()  # do anything with this?
```
Comment (fjetter) [inline distributed/scheduler.py:2703]:
```suggestion
        ts.waiters = None
```

**Comment by fjetter** [comment]
I think summarizing my review
- If possible we should use `assert foo is not None` instead of `if foo else ()` since it helps readability a lot. The scheduler is a messy, complex system and knowing for a fact that an attribute exists helps
- there are a bunch of places where we're clearing existing dicts/sets. In the spirit of only keeping what's necessary, we can instead just reset them to None
- There are a couple of very common sets that are used for every task and I think we should just keep them non-optional. This concerns primarily `dependencies` and `dependents`. They are all but guaranteed to be populated so we can initialize them right away and save ourselves a bit of complexity
- There are a couple of sets that are going to be populated for every single task but only for a certain amount of time. This includes `who_has`, `waiters`, `waiting_on`. We will only save ourselves memory by delaying initialization if we ensure that the sets are actually released once those collections are no longer required (which is typically true once a task is in memory). See the above comment about `clear`

File: distributed/scheduler.py (hunk 5 above)
Comment (milesgranger) [reply distributed/scheduler.py]: Good point https://github.com/dask/distributed/pull/8331/commits/62bf269536e9431c8097e4a5e99df124596a86db

File: distributed/scheduler.py (hunk 4 above)
Comment (milesgranger) [reply distributed/scheduler.py]: https://github.com/dask/distributed/pull/8331/commits/7df524092d3fd185da7d1be750a7be43cbf89430

File: distributed/scheduler.py [hunk 9]
```diff
@@ -762,7 +762,7 @@ def remove_from_processing(self, ts: TaskState) -> None:
         else:
             self._remove_from_task_prefix_count(ts)
         self.processing.remove(ts)
-        for dts in ts.dependencies:
+        for dts in ts.dependencies or ():
```
[the remark this answers is not in the document]
Comment (milesgranger) [reply distributed/scheduler.py]: https://github.com/dask/distributed/pull/8331/commits/7df524092d3fd185da7d1be750a7be43cbf89430

File: distributed/scheduler.py (hunk 2 above)
Comment (milesgranger) [reply distributed/scheduler.py]: https://github.com/dask/distributed/pull/8331/commits/8cd2b9e42374bfdc81d3b6fa2b34c0df4d6f0722

File: distributed/scheduler.py (hunk 3 above)
Comment (milesgranger) [reply distributed/scheduler.py]: https://github.com/dask/distributed/pull/8331/commits/7df524092d3fd185da7d1be750a7be43cbf89430

File: distributed/active_memory_manager.py [hunk 10]
```diff
@@ -309,19 +310,19 @@ def _find_dropper(
         def log_reject(msg: str) -> None:
             task_logger.debug("(drop, %s, %s) rejected: %s", ts, orig_candidates, msg)
 
-        if len(ts.who_has) - len(pending_drop) < 2:
+        if len(ts.who_has or ()) - len(pending_drop) < 2:
```
[the remark this answers is not in the document]
Comment (milesgranger) [reply distributed/active_memory_manager.py]: https://github.com/dask/distributed/pull/8331/commits/35d68f7d118077a93bd1904f549ea0d7f42da996

File: distributed/active_memory_manager.py [hunk 11]
```diff
@@ -533,7 +534,7 @@ def run(self) -> SuggestionGenerator:
         for ts in self.manager.scheduler.replicated_tasks:
             desired_replicas = 1  # TODO have a marker on TaskState
```
[the remark this answers is not in the document]
Comment (milesgranger) [reply distributed/active_memory_manager.py]: https://github.com/dask/distributed/pull/8331/commits/c847897e4158ee3707c44808e9b7561c603e007a

File: distributed/scheduler.py [hunk 12]
```diff
@@ -816,10 +821,10 @@ def _dec_needs_replica(self, ts: TaskState) -> None:
 
     def add_replica(self, ts: TaskState) -> None:
         """The worker acquired a replica of task"""
-        if self.scheduler.validate:
-            assert self not in ts.who_has
-            assert ts not in self.has_what
```
Comment (fjetter) [inline distributed/scheduler.py:821]: I changed add_replica to be idempotent so this validation is no longer valid

**Comment by fjetter** [comment]
test failures are all unrelated

File: distributed/active_memory_manager.py [hunk 13]
```diff
@@ -657,7 +661,7 @@ def run(self) -> SuggestionGenerator:
                 # would have stopped earlier
                 continue
 
-            if len(ts.who_has) > 1:
+            if ts.who_has:
```
Comment (fjetter) [inline distributed/active_memory_manager.py:664]: this is different logic
```suggestion
            if ts.who_has and len(ts.who_has) > 1:
```

**Comment by fjetter** [comment]
> test failures are all unrelated

kind of sad considering all the failures...

**Pull request merged**